### DISCUSSION TITLE
Train + color sensor - best practices

### DISCUSSION THREAD

**Discussion by adrianthomas** [pybricks/support#1978] [Q&A]
Hi all,

My 6 year old and I love the Pybricks concept and signed up to try out the color sensor smart train layout described here:
https://www.youtube.com/watch?v=kwnSbnlhOkA

We've been able to get it to work at very slow speeds (30 or less) and it definitely seems to work best with a green tile on the tracks.
Are there any "best practices" from other Lego train pybrickers to help us get started? 

- We've defined custom colors using the values coming off the sensor for red, green and yellow tiles
- The sensor is mounted at the front of the train, fairly close to the track
- The "print" debugging seems to show the color is seen

It seems as though the color sensor is detecting the color (as shown in the print output), but it's slow to respond

Any ideas or tips appreciated!

**Comment by laurensvalk** [comment]
Thanks for trying it out! Could you share a screenshot of your program? You can use the camera icon in the lower right to easily make a screenshot and just drag that into your reply here.

Did you add the track colors like in the video? It's important to add those so it can distinguish them from the actual colors.

**Reply by adrianthomas** [reply]
Thanks for getting back to us – yeah the other colours are added as well. We have reddish / brownish wooden floors, so I'm wondering if that is confusing it. 

I realise that a €15 sensor is not going to be perfect, but in your video it seems to stop almost instantly, so perhaps we've got something set up incorrectly.

Much appreciated!

![pybricks_program_2024-12-15T19_43_08](https://github.com/user-attachments/assets/efff9e9f-7ebb-431b-9dda-888cf9548501)

**Reply by laurensvalk** [reply]
Thanks for sharing that. The colors are looking good, i.e. properly distinct.

> It seems as though the color sensor is detecting the color (as shown in the print output), but it's slow to respond

When something is delayed it is always good to look at where your program is inserting delays, or making a test without such delays (in order to rule out delays you have no control over).

In your program, it seems that the first `2000 ms` wait is unintentional. This is where you start the motor. Then it drives for two whole seconds before even beginning to look for yellow. If it passed by yellow within those two seconds (which would be more likely at higher speed), it will have missed it.

> it definitely seems to work best with a green tile on the tracks

You don't have that  `2000 ms` wait after beginning to drive towards green, which could explain this part.

tl;dr, does it work better if you remove the first of your three 2-second waits?

----------------

Another way to experiment is to duplicate your program and keep the setup the same. But in your main program only begin to drive, wait for one color, and then stop. So no waiting for time at all, and no loops. See how fast you can drive and still properly detect the color and stop. Removing loops makes this sort of thing a lot easier to debug because you can follow along.

**Reply by BertLindeman** [reply]
Great experiment scenario, simplicity first, Laurens.
